server/session_map.cpp

```diff
@@ -0,0 +1,50 @@
+#include "session_map.h"
+#include "server_session.h"
+
+#include <iostream>
+
+namespace nibaserver {
+
+bool session_map::write(const std::string &name, std::string &&data) {
+    std::shared_lock lock{mutex_};
```

Comment:
If the shared_lock is acquired by one or more people, it will prevent other functions from acquiring the unique_lock. User login and the periodical cleanup job will be pending. The other direction faces the same problem. Suppose the periodical cleanup is running, does it mean nobody can send out messages?

Reply:
Yeah that's right. Not sure if there is a good alternative

Reply:
Not scalable, though. There will be other operations that require the interaction between two users. If any of them between any pair of users are ongoing, then no one can login/logout....

Reply:
The `ptr->write` here is also a none blocking call, it would be pretty quick